lib/logger.js

```diff
@@ -48,7 +48,7 @@ function logger(options) {
             winston.format.timestamp(),
             winston.format.colorize(),
             winston.format.printf((info) => {
-                return `${info.timestamp} - ${info.level}: ${info.message}` + (info.meta ? `\n${util.inspect(info.meta, false, 4)}` : '');
+                return `${info.timestamp} - ${info.level}: ${info.message}${info.meta ? `\n${util.inspect(info.meta, false, 4)}` : ''}${info.stack ? `\n${info.stack}` : ''}`;
```

Comment:
Not really sure why this is needed.  You can already log the calling stack in the passed-in message via something like:  `log.error('PAL Server: ${msg}\n${error.message}\n${error.stack}');` (changed the quotes, but you get where I'm going with it).

In general, we should _discourage_ writing the stack for error messages, as doing so rarely provides meaningful details, but it does use a lot more of our logging resources (which are very finite).  An effective logging strategy should really only write the stack in situations where the calling function might be unclear and it is not possible to convey it more effectively through the message itself.

Reply:
👍 agree

Reply:
@DrEnter  Perhaps we can put this feature behind a flag so it would be only enabled during local development. Our efforts to work on michonne was greatly hindered by this. I think having to go in and change every single call to `log.error` to include the stack trace is not a viable way of debugging.

Reply:
A couple of things:
1. You REALLY shouldn't be using so many stack traces.  There are usage patterns that would accomplish what you are suggesting, but frankly they shouldn't be necessary.  Stacks traces are overly verbose and are often unhelpful.  We certainly don't want them on every "error" call or equivalent as they will very quickly make the log output so noisy as to be unusable.
2. It should be noted that you are going to have to change almost every single call regardless of how you approach this:  Most error logging calls aren't passed the `Error` object directly like this would require, since they expect a formatted message.  The default formatter will handle that situation given the logic modified above, but it won't happen with "normal" use.  The default `Error` object messages don't give any context to the error so passing them directly to the log calls is bad form.

Reply:
1. Can you please provide some examples of possible alternatives?
I can't agree with the statement that stack traces are not useful. That would be an indication of a build process that suppress or obfuscates useful information. It might be the case on some applications where cnn-logger is used, but when it comes to Michonne, we are in the process of continuously improving the developer experience, which includes better stack traces as well.
We have real word examples from our efforts with harmonized template integration where errors without the stack traces were hopelessly useless (i.e.: something is undefined somewhere).
There are common ways for reducing noise, like setting flags for the level of the logging. I reckon if the issues comes down to the number of errors (and their stack traces) then something is seriously wrong with the application. Errors should not be part of the expected behaviour. Verbosity comes from info logging, not warnings or errors. 
2. Silencing any error is usually not a good idea, the best practice for JS is to wrap the original error with your custom error message. Passing an error object to the console also enables debug tools and IDEs to enhance the log with rich features.
As part of our continuous improvement, we can update logging everywhere as we go.

I see no reason why this feature can't be enabled if we keep the backward compatibility with a feature flag (e.g.: an env variable).

Reply:
Just putting this here, debugging without a stack can be really challenging. Below is an error I'm currently try to debug but the source of that error I cannot track down easily because of the missing stack. Would be really good if we could have an option to at least enable it when we need.

<img width="836" alt="Screenshot 2019-07-23 at 14 58 53" src="https://user-images.githubusercontent.com/31280676/61718600-3d161f80-ad5b-11e9-918f-6512b9ed1ee6.png">

Reply:
I would argue that the error message itself is the problem.  It appears that this is a `catch` of some type, but not enough context about where it is being caught has been provided.  What function/method is this a part of?  If this is a large wrapping catch, maybe including the error stack is appropriate, but that should be part of the error message.

What we do not want to do is to start pushing content decisions into the library that is used to render that content.

Reply:
Yes the error message on its own is not great, it shows the top most React component that failed to render but does not give info about any child components that failed to render. Only the stack can show this, after adding it temporarily I'm now able to see exactly where this error occurred.

![stack](https://user-images.githubusercontent.com/31280676/61798154-b62b7a80-ae20-11e9-8c1a-7b664781f221.png)

Without this trace I would have to guess which child component threw an error. What do you recommend we do here?

Reply:
Most libraries support adding stack information to the logging message. You just have to pass the exception or stack trace to the method/function when you log an error. Or, in the case of Python >= 3.2, you pass an extra flag `stack_info=True` to the logging call.

Swift
```
do {
    ...
} catch let error {
    let callStackSymbols = NSThread.callStackSymbols()
    NSLog(@"failed! %@", [callStackSymbols]);
}

```
Objective-C
@try {
    ...
} @catch (NSException *exception) {
    NSLog(@"failed! %@", [exception callStackSymbols]);
}
```

Java
```
try {
   ...
} catch (Exception e) {
    log.error("failed!", e);
}
```

Reply:
In theory, we can configure the logger anyway we want. In this case `lib/logger.js` is not the logger object, it's a factory function. It's responsible for creating an instance of the logger (Winston) and deciding what to log and how:

```
winston.format.printf((info) => {
    return `${info.timestamp} - ${info.level}: ${info.message}` + (info.meta ? `\n${util.inspect(info.meta, false, 4)}` : '');
})
```

To replicate the behaviour of the other loggers, we can instruct Winston to add the stack trace when an exception is logged:

```
logger.log('error message', new Error('failed'));
logger.log(new Error('failed'));
```

For that we need to add `winston.format.errors({ stack: true })` to:

```
options.console.format = winston.format.combine(
    winston.format.errors({ stack: true }),
    winston.format.splat(),
    ...
```

More info here:
https://github.com/winstonjs/winston/blob/master/examples/errors.js